### ISSUE TITLE
Correct way to build with Analytics?

### ISSUE THREAD

**Issue by orzilca** [prebid/Prebid.js#1379]
## Type of issue
Question / Possible Bug?

## Description
Its not clear how one should build prebid.js using gulp with Analytics.
I  have tried the following:
<pre>gulp build --adapters=custom-adaptors.json --analytics=analytics.json</pre>

where analytics.json is a file with an array:
<pre>["roxot"]</pre>

The file is being built, without the analytics adapters.
I've added some logs along the way and managed to see the analytics folder is incorrect:

<pre>
20:01:13] Error: ENOENT: no such file or directory, scandir '../analytics'
  at Error (native)
  at Object.fs.readdirSync (fs.js:856:18)
  at Object.module.exports.getAnalyticsSources.directoryContents.filter [as getAnalyticsSources] (/Applications/XAMPP/xamppfiles/htdocs/cortex-prebid/gulpHelpers.js:33:34)
  at Gulp.<anonymous> (/Applications/XAMPP/xamppfiles/htdocs/cortex-prebid/gulpfile.js:70:36)
  at module.exports (/Applications/XAMPP/xamppfiles/htdocs/cortex-prebid/node_modules/orchestrator/lib/runTask.js:34:7)
  at Gulp.Orchestrator._runTask (/Applications/XAMPP/xamppfiles/htdocs/cortex-prebid/node_modules/orchestrator/index.js:273:3)
  at Gulp.Orchestrator._runStep (/Applications/XAMPP/xamppfiles/htdocs/cortex-prebid/node_modules/orchestrator/index.js:214:10)
  at /Applications/XAMPP/xamppfiles/htdocs/cortex-prebid/node_modules/orchestrator/index.js:279:18
  at finish (/Applications/XAMPP/xamppfiles/htdocs/cortex-prebid/node_modules/orchestrator/lib/runTask.js:21:8)
  at /Applications/XAMPP/xamppfiles/htdocs/cortex-prebid/node_modules/orchestrator/lib/runTask.js:52:4
  at f (/Applications/XAMPP/xamppfiles/htdocs/cortex-prebid/node_modules/end-of-stream/node_modules/once/once.js:17:25)
  at Transform.onend (/Applications/XAMPP/xamppfiles/htdocs/cortex-prebid/node_modules/end-of-stream/index.js:31:18)
  at emitNone (events.js:72:20)
  at Transform.emit (events.js:166:7)
  at /Applications/XAMPP/xamppfiles/htdocs/cortex-prebid/node_modules/readable-stream/lib/_stream_readable.js:965:16
  at nextTickCallbackWith0Args (node.js:433:9)
  at process._tickDomainCallback (node.js:403:13)

</pre>

### Expected results
Prebid.js 0.25 compiled with several bid adapters and analytics adapters

### Actual results
Prebid.js 0.25 compiled only with bid adapters.

## Platform details

OSX
gulp 3.9.1
npm 3.10.10

**Comment by GLStephen** [comment]
We ran into this recently when building adapters for http://ww.PubWise.io (pubwiseanalytics in the Prebid project) and determined the issue. As best we can tell, the "--analytics" flag is a flag to indicate using a directory with plugins up a directory from the project. We are adding ["pubwiseanalytics"] to the analytics file in the project to build properly. This introduces some build automation challenges and we're evaluating a PR to fix it.

**Comment by orzilca** [comment]
If you set gulpfile.js:
```
var analyticsDirectory = './src/adapters/analytics';
```

And then try to build using --analytics flag you get:
</pre>
{ [Error: ENOTDIR: not a directory, stat 'src/adapters/analytics/AnalyticsAdapter.js/package.json']
  errno: -20,
  code: 'ENOTDIR',
  syscall: 'stat',
  path: 'src/adapters/analytics/AnalyticsAdapter.js/package.json' }
{ [Error: ENOTDIR: not a directory, stat 'src/adapters/analytics/appnexus.js/package.json']
  errno: -20,
  code: 'ENOTDIR',
  syscall: 'stat',
  path: 'src/adapters/analytics/appnexus.js/package.json' }
{ [Error: ENOTDIR: not a directory, stat 'src/adapters/analytics/example.js/package.json']
  errno: -20,
  code: 'ENOTDIR',
  syscall: 'stat',
  path: 'src/adapters/analytics/example.js/package.json' }
{ [Error: ENOTDIR: not a directory, stat 'src/adapters/analytics/example2.js/package.json']
  errno: -20,
  code: 'ENOTDIR',
  syscall: 'stat',
  path: 'src/adapters/analytics/example2.js/package.json' }
{ [Error: ENOTDIR: not a directory, stat 'src/adapters/analytics/ga.js/package.json']
  errno: -20,
  code: 'ENOTDIR',
  syscall: 'stat',
  path: 'src/adapters/analytics/ga.js/package.json' }
{ [Error: ENOTDIR: not a directory, stat 'src/adapters/analytics/pubwiseanalytics.js/package.json']
  errno: -20,
  code: 'ENOTDIR',
  syscall: 'stat',
  path: 'src/adapters/analytics/pubwiseanalytics.js/package.json' }
{ [Error: ENOTDIR: not a directory, stat 'src/adapters/analytics/pulsepoint.js/package.json']
  errno: -20,
  code: 'ENOTDIR',
  syscall: 'stat',
  path: 'src/adapters/analytics/pulsepoint.js/package.json' }
{ [Error: ENOTDIR: not a directory, stat 'src/adapters/analytics/roxot.js/package.json']
  errno: -20,
  code: 'ENOTDIR',
  syscall: 'stat',
  path: 'src/adapters/analytics/roxot.js/package.json' }
{ [Error: ENOTDIR: not a directory, stat 'src/adapters/analytics/sharethrough_analytics.js/package.json']
  errno: -20,
  code: 'ENOTDIR',
  syscall: 'stat',
  path: 'src/adapters/analytics/sharethrough_analytics.js/package.json' }
</pre>

The ```getAnalyticsSources()``` is incorrect (gulpHelpers.js).
Or I'm doing something wrong - it is not clear at all and documentation is not helping in this case..

**Comment by mkendall07** [comment]
hey guys. Sorry for the confusion. The way to build now is to include it in the `modules` switch since everything is now a module.
example:
```gulp build```  // build step
```gulp bundle --modules=rubiconBidAdapter,appnexusAnalyticsAdapter```  // bundle with these modules. 

this is explain here but maybe we need some updated docs as well. https://github.com/prebid/Prebid.js/pull/1177#issuecomment-308885425

**Comment by mkendall07** [comment]
@GLStephen I think the above should help with your use case?

**Comment by orzilca** [comment]
@mkendall07 this applies to 0.25 also?

**Comment by orzilca** [comment]
@mkendall07 I guess not :{
```
Task 'bundle' is not in your gulpfile
```

**Comment by mkendall07** [comment]
Right now it's just on master. We will be a doing a release shortly though that includes these changes.

**Comment by orzilca** [comment]
@mkendall07 So I cannot use roxot / pulsepoint analytics with Prebid 0.25?

**Comment by mkendall07** [comment]
@orzilca sorry neglected to see you commented 0.25.0 in the ticket. This works for me. Steps I took:
`git checkout tags/0.25.0`
`yarn install`
<modify `analytics.json` to include `["roxot"]`
`gulp build --adapters=adapters.json --analytics=analytics.json`

![image](https://user-images.githubusercontent.com/1870166/28186427-181ddbdc-67e9-11e7-9090-fe38a745699a.png)

**Comment by orzilca** [comment]
@mkendall07 
I've just tried it, completely from the beginning (cloning git repo).

Getting this error:
<pre>
[14:05:18] Error: ENOENT: no such file or directory, scandir '../analytics'
  at Error (native)
  at Object.fs.readdirSync (fs.js:856:18)
  at Object.module.exports.getAnalyticsSources.directoryContents.filter [as getAnalyticsSources] (/Applications/XAMPP/xamppfiles/htdocs/Prebid.js/gulpHelpers.js:33:34)
  at Gulp.<anonymous> (/Applications/XAMPP/xamppfiles/htdocs/Prebid.js/gulpfile.js:68:36)
  at module.exports (/Applications/XAMPP/xamppfiles/htdocs/Prebid.js/node_modules/orchestrator/lib/runTask.js:34:7)
  at Gulp.Orchestrator._runTask (/Applications/XAMPP/xamppfiles/htdocs/Prebid.js/node_modules/orchestrator/index.js:273:3)
  at Gulp.Orchestrator._runStep (/Applications/XAMPP/xamppfiles/htdocs/Prebid.js/node_modules/orchestrator/index.js:214:10)
  at /Applications/XAMPP/xamppfiles/htdocs/Prebid.js/node_modules/orchestrator/index.js:279:18
  at finish (/Applications/XAMPP/xamppfiles/htdocs/Prebid.js/node_modules/orchestrator/lib/runTask.js:21:8)
  at /Applications/XAMPP/xamppfiles/htdocs/Prebid.js/node_modules/orchestrator/lib/runTask.js:52:4
  at f (/Applications/XAMPP/xamppfiles/htdocs/Prebid.js/node_modules/end-of-stream/node_modules/once/once.js:17:25)
  at Transform.onend (/Applications/XAMPP/xamppfiles/htdocs/Prebid.js/node_modules/end-of-stream/index.js:31:18)
  at emitNone (events.js:72:20)
  at Transform.emit (events.js:166:7)
  at /Applications/XAMPP/xamppfiles/htdocs/Prebid.js/node_modules/readable-stream/lib/_stream_readable.js:965:16
  at nextTickCallbackWith0Args (node.js:433:9)
  at process._tickDomainCallback (node.js:403:13)
</pre>

**Comment by GLStephen** [comment]
@orzilca This is the error we were getting.  The '../analytics' directory is relative to the gulpfile (I believe). Which is why I made my comment earlier about it looking for a directory in the parent of the Prebid repository locally. I never tried creating it just to see if it worked after that.

**Comment by orzilca** [comment]
@GLStephen 
I see its defined as ```.../analytics/```
But its weird as the folder is inside ```src/adapters/analytics/```

I have tried that also but its not working either..

I think I'm missing something here..

**Comment by GLStephen** [comment]
@orzilca Yes, it's here:

https://github.com/prebid/Prebid.js/blob/0f8de0c1f6d7b1463b0e105015d4c43d7e28c8a1/gulpHelpers.js

```
/*
--
  | * Get source files for analytics subdirectories in top-level `analytics`
  | * directory adjacent to Prebid.js.
  | * Invoke with gulp <task> --analytics
  | * Returns an array of source files for inclusion in build process
  | */
  | getAnalyticsSources: function(directory) {
  | if (!argv.analytics) {return [];} // empty arrays won't affect a standard build
  |  
  | const directoryContents = fs.readdirSync(directory);
  | return directoryContents
  | .filter(file => isModuleDirectory(path.join(directory, file)))
  | .map(moduleDirectory => {
  | const module = require(path.join(directory, moduleDirectory, MANIFEST));
  | return path.join(directory, moduleDirectory, module.main);
  | });
  | },
```

Called from:

https://github.com/prebid/Prebid.js/blob/0e72b548d30984da9e505a634d3b7b5cab17c3ab/gulpfile.js

```
var analyticsDirectory = '../analytics';

gulp.task('webpack', ['clean'], function () {

  // change output filename if argument --tag given
  if (argv.tag && argv.tag.length) {
    webpackConfig.output.filename = 'prebid.' + argv.tag + '.js';
  }

  webpackConfig.devtool = null;

  var externalModules = helpers.getArgModules();

  const analyticsSources = helpers.getAnalyticsSources(analyticsDirectory);
  const moduleSources = helpers.getModulePaths(externalModules);

  return gulp.src([].concat(moduleSources, analyticsSources, 'src/prebid.js'))
    .pipe(helpers.nameModules(externalModules))
    .pipe(webpack(webpackConfig))
    .pipe(replace('$prebid.version$', prebid.version))
    .pipe(uglify())
    .pipe(gulpif(file => file.basename === 'prebid.js', header(banner, { prebid: prebid })))
    .pipe(optimizejs())
    .pipe(gulp.dest('build/dist'))
    .pipe(connect.reload());
});
```

**Comment by orzilca** [comment]
@GLStephen 

I have moved the analytics folders to upper level, seems to be working now.

But why is it like that?
What is the purpose of having to place the folder manually on an upper level folder?
Doesn't make that much sense.. no?

**Comment by GLStephen** [comment]
@orzilca No, on some level it's not right. I'm looking into a PR, but it's working fine, or as expected, for @mkendall07. If we can work that apart and figure out what the disconnect is, then we'll have an idea of where to go next.

**Comment by mkendall07** [comment]
hey guys - this might be a bug in 0.25.0. We have an adapter that can be built from `../analytics` and that exists on my file system. We probably won't fix this though because 0.26.0 will be out in a few days and the build system is different now.

**Comment by mkendall07** [comment]
Closing this out as 0.26.x is released and we have a path forward for building any type of module from command line interface. Please do let me know if you have trouble with the latest build code.
